[SPARK-19160][PYTHON][SQL] Add udf decorator

zero323 opened the pull request:
## What changes were proposed in this pull request?
    
This PR adds `udf` decorator syntax as proposed in [SPARK-19160](https://issues.apache.org/jira/browse/SPARK-19160).
    
This allows users to define UDF using simplified syntax:
    
```python
from pyspark.sql.decorators import udf
    
@udf(IntegerType())
def add_one(x):
    """Adds one"""
    if x is not None:
        return x + 1
 ```
    
without need to define a separate function and udf.


## How was this patch tested?

Existing unit tests to ensure backward compatibility and additional unit tests covering new functionality.

zero323 committed:
Add udf decorator

This PR adds `udf` decorator syntax as proposed in [SPARK-19160](https://issues.apache.org/jira/browse/SPARK-19160).

This allows users to define UDF using simplified syntax:

```
from pyspark.sql.decorators import udf

@udf(IntegerType())
def add_one(x):
    """Adds one"""
    if x is not None:
        return x + 1
```

without need to define a separate function and udf.

Tested wiht existing unit tests to ensure backward compatibility and additional unit tests covering new functionality.

SparkQA commented:
**[Test build #71158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71158/testReport)** for PR 16533 at commit [`8645ed8`](https://github.com/apache/spark/commit/8645ed82fef9f1d5bebdc8793e92da105d47a151).

SparkQA commented:
**[Test build #71158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71158/testReport)** for PR 16533 at commit [`8645ed8`](https://github.com/apache/spark/commit/8645ed82fef9f1d5bebdc8793e92da105d47a151).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71158/
Test PASSed.

rdblue commented on python/pyspark/sql/decorators.py:
```diff
@@ -0,0 +1,25 @@
+from pyspark.sql.functions import udf as _udf
+
+
+def udf(returnType=StringType()):
```
I don't think this should conflict with pyspark.sql.functions.udf. If this were defined, then it would break existing code that uses functions.udf. It may be possible to have a decorator definition that maintains that API and acts as a decorator, but otherwise this should be renamed to something like `register_udf`.

rdblue replied:
Can you also add docstrings for this?

zero323 replied:
Regarding name clashes I would like to keep it as a separated module and let user decide if prefer to import from `sql.functions` or `sql.decorators`  though technically speaking, with "parenthless" version, we could just merge it with `pyspark.sql.functions.udf`. The problem is it becomes rather ugly to document because the argument can be either `DataType` or `Callable`. 

This is why I wait with adding documentation but I pushed some doctests.

zero323 replied:
Of course it could be just moved to `sql.functions` and renamed explicitly as `udf_decorator`. Then user can use an alias to get shorter call:

    from sql.functions import udf_decorator as udf

rdblue replied:
I think that the two should be merged. If they have the same name, then what works in one place breaks in others and the two are mutually exclusive. Our users share code quite a bit and it is common to copy & paste UDFs. That should be fine, or should result in "NameError: name 'udf' is not defined", rather than having different behavior.

With the changes to the decorator to support using it without parens, it's already close to being compatible with the current `functions.udf`.

zero323 replied:
OK, I moved decorator to `sql.functions` as `udf_decorator`. _Explicit is better than implicit_, right? Nevertheless I am reluctant to further attempts to merge it with the main `udf` function.

Technically It is possible, but it will require all kinds of special cases. It would be manageable for now, when we only support `returnType`, but if we ever decide to add further arguments (like `name`) it would be a mess. Especially when project is committed to supporting legacy Python versions. Otherwise we could just use `keyword` only arguments:

    def udf(func=None, *, returnType=StringType()):
        ... 

We could mimic that to some extent, without using backwards compatibility, by using keyword arguments:

    def udf(func=None, **kwargs):
        returnType = kwargs.get("returnType", StringType())
        ...

but I don't think it is worth loosing explicit argument list.

rdblue commented on python/pyspark/sql/tests.py:
```diff
@@ -429,6 +429,17 @@ def test_udf_with_input_file_name(self):
+        from pyspark.sql.decorators import udf
+        from pyspark.sql.types import IntegerType
+
+        @udf(IntegerType())
```
Could you add a test case for the default return type? I think it currently requires empty-parens:

```
@udf()
def truncate(s):
    return s[:10]
```

zero323 replied:
It does. In general decorators with arguments should be called with empty parens. This call is equivalent to:

    udf()(truncate)

We can support parentheses-less call but it requires a bit ugly checks on runtime. I'll push an example implementation and additional tests in a moment.

rdblue commented:
I think if the `udf` method takes as its first arg `func=None`, then you can match the old API and take care of the empty-paren problem. If a callable isn't passed as the first arg, then return a decorator function. If a callable is passed in, then register it. You can do something similar with `*args` if you don't want to require passing `returnType` as a named arg.

SparkQA commented:
**[Test build #71201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71201/testReport)** for PR 16533 at commit [`8daa362`](https://github.com/apache/spark/commit/8daa36287922f46f23ff164a8ba354bce9e4ba18).

SparkQA commented:
**[Test build #71201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71201/testReport)** for PR 16533 at commit [`8daa362`](https://github.com/apache/spark/commit/8daa36287922f46f23ff164a8ba354bce9e4ba18).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71201/
Test PASSed.

SparkQA commented:
**[Test build #71202 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71202/testReport)** for PR 16533 at commit [`d35cb06`](https://github.com/apache/spark/commit/d35cb06f7d63e8bdfb4fdc040cf204753b4711be).

SparkQA commented:
**[Test build #71202 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71202/testReport)** for PR 16533 at commit [`d35cb06`](https://github.com/apache/spark/commit/d35cb06f7d63e8bdfb4fdc040cf204753b4711be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71202/
Test PASSed.

SparkQA commented:
**[Test build #71227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71227/testReport)** for PR 16533 at commit [`8efc00b`](https://github.com/apache/spark/commit/8efc00b3ea1a6a8b842b204be258ef4d927a14d3).

SparkQA commented:
**[Test build #71227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71227/testReport)** for PR 16533 at commit [`8efc00b`](https://github.com/apache/spark/commit/8efc00b3ea1a6a8b842b204be258ef4d927a14d3).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71227/
Test FAILed.

SparkQA commented:
**[Test build #71228 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71228/testReport)** for PR 16533 at commit [`d96a63e`](https://github.com/apache/spark/commit/d96a63e74fdae2579b49ed8470639e608cd46ab5).

SparkQA commented:
**[Test build #71228 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71228/testReport)** for PR 16533 at commit [`d96a63e`](https://github.com/apache/spark/commit/d96a63e74fdae2579b49ed8470639e608cd46ab5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71228/
Test PASSed.

zero323 commented:
@rdblue  Unified `udf` which supports all cases (`udf(f)`, decorator with parentheses and without, decorator with `returnType`) could be implemented like this: 

```python
def _udf(f, returnType=StringType()):
    """Creates a :class:`Column` expression representing a user defined function (UDF).

    .. note:: The user-defined functions must be deterministic. Due to optimization,
        duplicate invocations may be eliminated or the function may even be invoked more times than
        it is present in the query.

    :param f: python function
    :param returnType: a :class:`pyspark.sql.types.DataType` object

    >>> from pyspark.sql.types import IntegerType
    >>> slen = udf(lambda s: len(s), IntegerType())
    >>> df.select(slen(df.name).alias('slen')).collect()
    [Row(slen=5), Row(slen=3)]
    >>> @udf
    ... def to_upper(s):
    ...     if s is not None:
    ...         return s.upper()
    ...
    >>> @udf(returnType=IntegerType())
    ... def add_one(x):
    ...     if x is not None:
    ...         return x + 1
    ...
    """
    return UserDefinedFunction(f, returnType)


@functools.wraps(_udf)
def udf(f=None, **kwargs):
    """A decorator version of pyspark.sql.functions.udf
    """
    returnType = kwargs.get("returnType", StringType())

    if f is None:
        return functools.partial(_udf, returnType=returnType)
    else:
        return _udf(f=f, returnType=returnType)
```

We can use `functools.wraps` to update docstring. There are two problems with this design:

- [Once again](https://github.com/apache/spark/pull/16534#issuecomment-271735937) in legacy Python we won't get meaningful argument list.
- It will break pre-existing code which consider `returnType` to be a positional argument (including a bunch of tests).

Do you think this is an acceptable trade-off?

rdblue commented:
@zero323, I think that the decorator and existing UDF factory method should be the same, but that we can't break existing code. Can you explain why this necessarily breaks code that relies on returnType as a positional arg? What about something like this:

```python
@functools.wraps(_udf)
def udf(f=None, returnType=StringType()):
    """A decorator version of pyspark.sql.functions.udf
    """
    if f is None:
        return functools.partial(_udf, returnType=returnType)
    else:
        return _udf(f=f, returnType=returnType)
```

zero323 commented:
@rdblue Good question. I vaguely remember I had some motivation to avoid this but now I cannot recall why.  In general I really liked the initial approach because it allowed us to write:

```python
@udf(IntegerType())
def identity(x):
    return x
```
instead of 

```python
@udf(returnType=IntegerType())
def identity(x):
    return x
```

There is another trick, which some libraries that depend heavily on decorators, use:

```python 
@functools.wraps(_udf)
def udf(f=None, returnType=StringType()):
    """A decorator version of pyspark.sql.functions.udf
    """
    if f is None or isinstance(f, DataType):
        return functools.partial(_udf, returnType=returnType)
    else:
        return _udf(f=f, returnType=returnType)
```

but it smells. 

I think that:

```python
@functools.wraps(_udf)
def udf(f=None, returnType=StringType()):
    """A decorator version of pyspark.sql.functions.udf
    """
    if  isinstance(f, DataType):
       raise TypeError("returnType with decorator should be provided as a keyword argument")
    if f is None:
        return functools.partial(_udf, returnType=returnType)
    else:
        return _udf(f=f, returnType=returnType)
```
could be an acceptable trade-off. It doesn't brake current API and clearly communicates possible issues.

SparkQA commented:
**[Test build #71678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71678/testReport)** for PR 16533 at commit [`88326fd`](https://github.com/apache/spark/commit/88326fd548ffcfddccc47d04eca21c35a5ab5562).

SparkQA commented:
**[Test build #71678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71678/testReport)** for PR 16533 at commit [`88326fd`](https://github.com/apache/spark/commit/88326fd548ffcfddccc47d04eca21c35a5ab5562).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71678/
Test FAILed.

rdblue commented:
Yeah, I added the `isinstance(f, DataType)` trick to my local tests. I'd add it if it were up to me, but I'm fine requiring `returnType` to be a keyword arg if you feel strongly about it.

Looks like tests are broken, maybe rebase?

SparkQA commented:
**[Test build #71682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71682/testReport)** for PR 16533 at commit [`4ed5f6f`](https://github.com/apache/spark/commit/4ed5f6f203b69508829f4f396c52a5a789c88e8f).

SparkQA commented:
**[Test build #71682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71682/testReport)** for PR 16533 at commit [`4ed5f6f`](https://github.com/apache/spark/commit/4ed5f6f203b69508829f4f396c52a5a789c88e8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71682/
Test PASSed.

SparkQA commented:
**[Test build #71721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71721/testReport)** for PR 16533 at commit [`f55c6ee`](https://github.com/apache/spark/commit/f55c6ee80f550f8e7dae25e0ced2cd2575e34551).

zero323 commented:
@rdblue You're probably right and I am overthinking this. After all the goal here is to make this as user friendly as possible, and while `isinstance` is not pretty it does the trick.  If we ever decide to add more arguments these can be made keyword only to avoid ambiguity. 

Rebased, squashed and tested. Thank you for your support and advice!

SparkQA commented:
**[Test build #71721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71721/testReport)** for PR 16533 at commit [`f55c6ee`](https://github.com/apache/spark/commit/f55c6ee80f550f8e7dae25e0ced2cd2575e34551).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71721/
Test PASSed.

rdblue commented:
+1

Looks good to me. Thanks, @zero323!

holdenk commented on python/pyspark/sql/functions.py:
```diff
@@ -1847,8 +1848,7 @@ def __call__(self, *cols):
+def _udf(f, returnType=StringType()):
     """Creates a :class:`Column` expression representing a user defined function (UDF).
 
     .. note:: The user-defined functions must be deterministic. Due to optimization,
```
Even though we've got this with an _ maybe through in the docstring its for internal use only? Also since this is for internal use only maybe most of this should be moved to the public udf function now instead of on the internal only method?

Unless we intend for users to call `_udf` in which case we need to give it a different name.

zero323 replied:
If we don't mind some discrepancy between docstring and usage (like I mentioned I believe that documenting internals with switching between decorator with partial and full application won't do us any good here) it can be pushed as an internal helper into `udf`.

holdenk commented on python/pyspark/sql/functions.py:
```diff
@@ -1862,9 +1862,33 @@ def udf(f, returnType=StringType()):
     >>> slen = udf(lambda s: len(s), IntegerType())
     >>> df.select(slen(df.name).alias('slen')).collect()
     [Row(slen=5), Row(slen=3)]
+    >>> @udf
```
I like this example as a doctest - but maybe good to also show these being applied in the doctest.

holdenk commented on python/pyspark/sql/functions.py:
```diff
@@ -1862,9 +1862,33 @@ def udf(f, returnType=StringType()):
 
+
+@functools.wraps(_udf)
+def udf(f=None, returnType=StringType()):
```
Should we keep the since annotation? Also maybe explain what these params are in the docstring (which can be based on the docstirng you currently have for _udf but updated).

zero323 replied:
We can, but I am not sure if we should keep the original version or acknowledge change of behavior and use potential merge version.

Regarding docstring I am a bit reluctant. It can be documented for dev purposes but any user facing doc will have to expose ugly implementation details. 

    def udf(
        f: typing.Union[None, typing.Callable, pyspark.sql.types.DataType],  
        returnType: pyspark.sql.types.DataType=StringType()
     ) -> Union[types.FunctionType, functools.partial]:
        ...

Maybe I a wrong but it can bring more confusion than it is worth. We can of course move `_udf` docstring and skip wrapping (we considered different implementations where signatures differed) but of course it won't be accurate.

holdenk commented on python/pyspark/sql/functions.py:
```diff
@@ -1862,9 +1862,33 @@ def udf(f, returnType=StringType()):
+def udf(f=None, returnType=StringType()):
+    """A decorator version of pyspark.sql.functions.udf
+    """
+    if f is None or isinstance(f, DataType):
```
Maybe you can mention why this would be the case in the comment as well?

zero323 replied:
Do you mean something like this?

```python
@functools.wraps(_udf)
def udf(f=None, returnType=StringType()):
    """A decorator version of pyspark.sql.functions.udf
    """
    if (
            # Decorator as @udf() or @udf
            f is None
            # Decorator with positional DataType @udf(IntegerType())
            or isinstance(f, DataType)
    ):
    ...
```

holdenk reviewed (commented):
First quick skim, looks like a potentially useful feature - lets see where this goes. I'll do a more detailed pass after Spark Summit East :)

SparkQA commented:
**[Test build #72072 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72072/testReport)** for PR 16533 at commit [`2906889`](https://github.com/apache/spark/commit/2906889b6acdcb90812204c67241d57c49165da5).

SparkQA commented:
**[Test build #72072 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72072/testReport)** for PR 16533 at commit [`2906889`](https://github.com/apache/spark/commit/2906889b6acdcb90812204c67241d57c49165da5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72072/
Test FAILed.

SparkQA commented:
**[Test build #72073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72073/testReport)** for PR 16533 at commit [`218a800`](https://github.com/apache/spark/commit/218a800f37b34247576189b119cd7d5d2875f71c).

SparkQA commented:
**[Test build #72073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72073/testReport)** for PR 16533 at commit [`218a800`](https://github.com/apache/spark/commit/218a800f37b34247576189b119cd7d5d2875f71c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72073/
Test FAILed.

SparkQA commented:
**[Test build #72074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72074/testReport)** for PR 16533 at commit [`d8f7f44`](https://github.com/apache/spark/commit/d8f7f444db73a3da1ce697ef72b2dfdcc3b9f112).

SparkQA commented:
**[Test build #72074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72074/testReport)** for PR 16533 at commit [`d8f7f44`](https://github.com/apache/spark/commit/d8f7f444db73a3da1ce697ef72b2dfdcc3b9f112).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72074/
Test PASSed.

rxin commented:
Can return type also take a string?

zero323 commented:
@rxin Sure. Give me a second. In a meantime would consider reviewing #16742 ([SPARK-19403](https://issues.apache.org/jira/browse/SPARK-19403))?

lins05 commented:
What about also support using the type name without the parentheses, as a syntax sugar? e.g.:

```python
@udf(returnType =IntegerType) # instead of IntegerType()
def f(...):
   ...
```

It can be achieved by checking if the returnType is a subclass of DataType:

```python
if isinstance(returnType, type) and DataType in returnType.__mro__:
   returnType = returnType()
else ...
```

zero323 commented:
@rxin 

We can do something like this:

```python
class UserDefinedFunction(object):
    """
    User defined function in Python

    .. versionadded:: 1.3
    """
    def __init__(self, func, returnType, name=None):
        self.func = func
        self.returnType = returnType
        self._judf = self._create_judf(name)

    def _jdt(self, spark):
        if isinstance(self.returnType, DataType):
            return spark._jsparksession.parsedatatype(self.returntype.json())
        else:
            jvm = spark.sparkContext._jvm
            return jvm.org.apache.spark.sql.catalyst.parser.CatalystSqlParser.parseDataType(self.returnType)

    def _create_judf(self, name):
        from pyspark.sql import SparkSession
        sc = SparkContext.getOrCreate()
        wrapped_func = _wrap_function(sc, self.func, self.returnType)
        spark = SparkSession.builder.getOrCreate()
        jdt = self._jdt(spark)
        if name is None:
            f = self.func
            name = f.__name__ if hasattr(f, '__name__') else f.__class__.__name__
        judf = sc._jvm.org.apache.spark.sql.execution.python.UserDefinedPythonFunction(
            name, wrapped_func, jdt)
        return judf

    def __call__(self, *cols):
        sc = SparkContext._active_spark_context
        jc = self._judf.apply(_to_seq(sc, cols, _to_java_column))
        return Column(jc)
```

and then easily support:

```python
@udf("integer")
def f(x):
    return x
```
but I am not sure if this is the right place to handle this. For example it will raise some questions what should be the type of `UserDefinedFunction.returnType`. `DataType` or the thing we passed (`DataType` or `str`)?  I am happy to work on this but I would prefer to add it as a separate task to [SPARK-19159](https://issues.apache.org/jira/browse/SPARK-19159). What do you think?

zero323 commented:
@lins05  I am not convinced. This has fairly limited  applications (we cannot use this to define complex types) and could be confusing for users (why we can use it here but not for `Column.cast`, StructField.__init__` and so on).

rdblue commented:
I agree with @zero323. I don't think this method should change how users interact with DataTypes. There is a precedent for using type strings instead, as @rxin suggested, and that's a good idea that fits the need better than allowing users to pass the type class instead of an instance.

SparkQA commented:
**[Test build #72241 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72241/testReport)** for PR 16533 at commit [`8a2f1bb`](https://github.com/apache/spark/commit/8a2f1bb29a9a18f25fb6a1662f7190498c3e17ff).

SparkQA commented:
**[Test build #72241 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72241/testReport)** for PR 16533 at commit [`8a2f1bb`](https://github.com/apache/spark/commit/8a2f1bb29a9a18f25fb6a1662f7190498c3e17ff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72241/
Test PASSed.

zero323 commented:
Handling data type strings proved to be much simpler than I thought (I missed that PySpark already has its own parser): [SPARK-19427](https://issues.apache.org/jira/browse/SPARK-19427) and #16769

SparkQA commented:
**[Test build #72328 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72328/testReport)** for PR 16533 at commit [`cd0af7c`](https://github.com/apache/spark/commit/cd0af7cd81b320951090776c7339b89f495650a6).

SparkQA commented:
**[Test build #72328 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72328/testReport)** for PR 16533 at commit [`cd0af7c`](https://github.com/apache/spark/commit/cd0af7cd81b320951090776c7339b89f495650a6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72328/
Test PASSed.

SparkQA commented:
**[Test build #72331 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72331/testReport)** for PR 16533 at commit [`7dc670d`](https://github.com/apache/spark/commit/7dc670deae2667d4f64aeed78b77fc4f41df9da7).

zero323 commented:
@holdenk If believe this is a good as it gets at this moment.

SparkQA commented:
**[Test build #72331 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72331/testReport)** for PR 16533 at commit [`7dc670d`](https://github.com/apache/spark/commit/7dc670deae2667d4f64aeed78b77fc4f41df9da7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72331/
Test PASSed.

holdenk commented:
Ok back from the summit, let me take a look through - sorry for the slowness with my travel.

holdenk commented:
So one minor thing, I don't see support for taking the return type as string - was that decided to be dropped?

zero323 commented:
@holdenk  We cannot support it without modifying `UserDefinedFunction` first so I decided to open a separate PR for this: https://github.com/apache/spark/pull/16769

holdenk commented:
Great, I've merged that PR, can you update the instance checks here as required and add a test using the decorator with a string?

SparkQA commented:
**[Test build #72821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72821/testReport)** for PR 16533 at commit [`f9b1eb2`](https://github.com/apache/spark/commit/f9b1eb2bd1dfaa40605e62beb47bdcdda9101d1a).

zero323 committed:
Add tests for udf decorator with str returnType

SparkQA commented:
**[Test build #72822 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72822/testReport)** for PR 16533 at commit [`cafdf44`](https://github.com/apache/spark/commit/cafdf442c083f524bb9942f301f35fc0b2b4290f).

SparkQA commented:
**[Test build #72822 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72822/testReport)** for PR 16533 at commit [`cafdf44`](https://github.com/apache/spark/commit/cafdf442c083f524bb9942f301f35fc0b2b4290f).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72822/
Test FAILed.

SparkQA commented:
**[Test build #72823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72823/testReport)** for PR 16533 at commit [`4f5c643`](https://github.com/apache/spark/commit/4f5c64324306b6daf3660ef075223edd0460ba11).

SparkQA commented:
**[Test build #72824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72824/testReport)** for PR 16533 at commit [`01be3ec`](https://github.com/apache/spark/commit/01be3ece4b8fe7ad6c8e312e38c0b0570673fe14).

SparkQA commented:
**[Test build #72821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72821/testReport)** for PR 16533 at commit [`f9b1eb2`](https://github.com/apache/spark/commit/f9b1eb2bd1dfaa40605e62beb47bdcdda9101d1a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72821/
Test PASSed.

holdenk commented:
Ok this looks pretty reasonable, but I'm going to give it a day incase anyone (e.g. @rdblue or @HyukjinKwon ) have any last comments or suggestions.

Thanks so much for working on this @zero323 its going to really improve the simplicity of writing UDFs for Python users :)

SparkQA commented:
**[Test build #72823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72823/testReport)** for PR 16533 at commit [`4f5c643`](https://github.com/apache/spark/commit/4f5c64324306b6daf3660ef075223edd0460ba11).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72823/
Test FAILed.

SparkQA commented:
**[Test build #72825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72825/testReport)** for PR 16533 at commit [`282e943`](https://github.com/apache/spark/commit/282e94395208187c8590fa11a0d1a90af659eb8b).

SparkQA commented:
**[Test build #72824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72824/testReport)** for PR 16533 at commit [`01be3ec`](https://github.com/apache/spark/commit/01be3ece4b8fe7ad6c8e312e38c0b0570673fe14).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72824/
Test FAILed.

SparkQA commented:
**[Test build #72825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72825/testReport)** for PR 16533 at commit [`282e943`](https://github.com/apache/spark/commit/282e94395208187c8590fa11a0d1a90af659eb8b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72825/
Test PASSed.

HyukjinKwon commented:
Thanks for cc'ing me. I like this pythonic way. +1 and looks okay to me too despite of the trick in argument checking.

holdenk commented:
Great :) So the sql python tests have changed a bit which means this needs another update (sorry @zero323 ) give me a ping when its updated and I'll try and come back to this quickly :)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #72885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72885/testReport)** for PR 16533 at commit [`8f8d39e`](https://github.com/apache/spark/commit/8f8d39e11e18eeff6096b196353d9a0427fd92e1).

SparkQA commented:
**[Test build #72885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72885/testReport)** for PR 16533 at commit [`8f8d39e`](https://github.com/apache/spark/commit/8f8d39e11e18eeff6096b196353d9a0427fd92e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72885/
Test PASSed.

zero323 commented:
@holdenk done.

holdenk commented:
Thanks for working on this @zero323, merged to master :)

zero323 commented:
Thanks @holdenk
